File: 19_数学/因数筛/莫比乌斯反演/template-集合变换.go
// 集合上的zeta/mobius变换：高维前缀和(SOSDp)

package main

// 超集的前缀和变换.
//
// c2[v] = c1[s1] + c1[s2] + ... + c1[s_n] (si & v == v)
func SupersetZeta(c1 []int) {
	n := len(c1)
	if n&(n-1) != 0 {
		panic("n must be a power of 2")
	}
	for i := 1; i < n; i <<= 1 {
		for j := 0; j < n; j++ {
			if j&i == 0 {
				c1[j] += c1[j|i]
			}
		}
	}
}

// 超集的前缀和逆变换.
func SupersetMoebius(c2 []int) {
	n := len(c2)
	if n&(n-1) != 0 {
		panic("n must be a power of 2")
	}
	for i := 1; i < n; i <<= 1 {
		for j := 0; j < n; j++ {
			if j&i == 0 {
				c2[j] -= c2[j|i]
			}
		}
	}
}

// 子集的前缀和变换.
//
// c2[v] = c1[s1] + c1[s2] + ... + c1[v] (v & si == si)
func SubsetZeta(c1 []int) {
	n := len(c1)
	if n&(n-1) != 0 {
		panic("n must be a power of 2")
	}
	for i := 1; i < n; i <<= 1 {
		for j := 0; j < n; j++ {
			if j&i == 0 {
				c1[j|i] += c1[j]
			}
		}
	}
}

// 子集的前缀和逆变换.
func SubsetMoebius(c2 []int) {
	n := len(c2)
	if n&(n-1) != 0 {
		panic("n must be a power of 2")
	}
	for i := 1; i < n; i <<= 1 {
		for j := 0; j < n; j++ {
			if j&i == 0 {
				c2[j|i] -= c2[j]
			}
		}
	}
}

File: 19_数学/因数筛/莫比乌斯反演/template-集合变换.py
from typing import List


def supersetZeta(c1: List[int]):
    """超集的前缀和变换.

    c2[v] = c1[s1] + c1[s2] + ... + c1[s_n] (si & v == v)
    """
    n = len(c1)
    assert n & (n - 1) == 0, "n must be a power of 2"
    i = 1
    while i < n:
        for j in range(n):
            if not j & i:
                c1[j] += c1[j | i]
        i <<= 1


def supersetMobius(c2: List[int]):
    """超集的前缀和逆变换."""
    n = len(c2)
    assert n & (n - 1) == 0, "n must be a power of 2"
    i = 1
    while i < n:
        for j in range(n):
            if not j & i:
                c2[j] -= c2[j | i]
        i <<= 1


def subsetZeta(c1: List[int]):
    """子集的前缀和变换.
    c2[v] = c1[s1] + c1[s2] + ... + c1[v] (v & si == si)
    """
    n = len(c1)
    assert n & (n - 1) == 0, "n must be a power of 2"
    i = 1
    while i < n:
        for j in range(n):
            if not j & i:
                c1[j | i] += c1[j]
        i <<= 1


def subsetMobius(c2: List[int]):
    """子集的前缀和逆变换."""
    n = len(c2)
    assert n & (n - 1) == 0, "n must be a power of 2"
    i = 1
    while i < n:
        for j in range(n):
            if not j & i:
                c2[j | i] -= c2[j]
        i <<= 1


if __name__ == "__main__":
    nums = [1, 2, 3, 4]
    subsetSum = [0] * (1 << len(nums))
    for i in range(len(nums)):
        subsetSum[1 << i] = nums[i]
    subsetZeta(subsetSum)
    print(subsetSum)
